cmd/govim/fillstruct.go

```diff
@@ -34,33 +35,112 @@ func (v *vimstate) fillStruct(flags govim.CommandFlags, args ...string) error {
 		return nil
 	}
 
-	buri := b.URI()
-	var edits []protocol.TextEdit
-	for _, ca := range codeActions {
-		// there should be just a single file
-		dcs := ca.Edit.DocumentChanges
-		switch len(dcs) {
-		case 1:
-			dc := dcs[0]
-			// verify that the URI and version of the edits match the buffer
-			euri := span.URI(dc.TextDocument.TextDocumentIdentifier.URI)
-			if euri != buri {
-				return fmt.Errorf("got edits for file %v, but buffer is %v", euri, buri)
+	// TODO: revisit this logic when gopls enables us to distingush two different
+	// code action responses.
+	//
+	// The response from gopls contain Commands, but gopls currently responds
+	// with all code actions at the current line (instead of the exact range
+	// passed as parameter).
+	//
+	// If there are more than one command returned, we can't apply them all since
+	// each one of them contains unspecified parameters that are bound to current
+	// version of the document.
+
+	// The gopls ExecuteCommand is blocking, and gopls will call back to govim
+	// using ApplyEdit that must be handled before the blocking is released.
+	// Since fillstruct is ordered by the user (and the single threaded nature
+	// of vim), we are effectively blocking ApplyEdit from modifying buffers.
+	//
+	// To prevent a deadlock, we create a channel that ApplyEdit can use pass
+	// edits to this thread (if needed). And then call ExecuteCommand in a
+	// separate goroutine so that this thread can go on updating buffers
+	// until the ExecuteCommand is released. When it is, we implicitly know
+	// that ApplyEdit has been processed.
+	editsCh := make(chan applyEditCall)
+	v.govimplugin.applyEditsLock.Lock()
+	v.govimplugin.applyEditsCh = editsCh
+	v.govimplugin.applyEditsLock.Unlock()
+	done := make(chan struct{})
+
+	v.tomb.Go(func() error {
+		// We can only apply one command at the moment since they all target the same document
+		// version. Let's go for the first one and let the user call fillstruct again if they
+		// want to fill several structs on the same line.
+		ca := codeActions[0]
```

Comment:
I opted this mainly because multiple responses are an edge case, and this code block is a temporary thing until https://github.com/golang/go/issues/40438 is fixed. Since we pass `Only: [..] protocol.RefactorRewrite` we will only get `fillstruct` responses (afaik that's the only supported rewrite action for .go files).

What would the purpose of selecting by title be? If it is to ensure that the response is a `fillstruct` command we could instead look at the `Command` field within the command in the response. Example response:
```
Command:     &protocol.Command{
    Title:     "Fill foo with default values",
    Command:   "fill_struct",
    Arguments: {
        "{\"uri\":\"file:///private/var/folders/xm/ls208xd174v95pgd20rn_qbh0000gn/T/tmp.lmXRD9mP/main.go\",\"range\":{\"start\":{\"line\":7,\"character\":6},\"end\":{\"line\":7,\"character\":6}}}",
    },
},
```

If it is to ensure that we pick the "correct one" that has a matching name, we must then first extract the struct identifier in `govim`. I avoided that complexity in this PR since it would only be needed when there are multiple empty structs to fill on the same line.